pkg/virt-controller/services/template.go

```diff
@@ -19,7 +21,27 @@ func (t *templateService) RenderLaunchManifest(vm *v1.VM) (*kubev1.Pod, error) {
 	precond.MustNotBeNil(vm)
 	domain := precond.MustNotBeEmpty(vm.GetObjectMeta().GetName())
 	uid := precond.MustNotBeEmpty(string(vm.GetObjectMeta().GetUID()))
-	True := true
+
+	// VM target container
+	container := kubev1.Container{
+		Name:            "compute",
+		Image:           t.launcherImage,
+		ImagePullPolicy: kubev1.PullIfNotPresent,
+		Command:         []string{"/virt-launcher", "-qemu-timeout", "60s"},
+	}
+
+	// Set up spice ports
+	ports := []kubev1.ContainerPort{}
+	for i, g := range vm.Spec.Domain.Devices.Graphics {
+		if strings.ToLower(g.Type) == "spice" {
+			ports = append(ports, kubev1.ContainerPort{
+				ContainerPort: g.Port,
+				Name:          "spice" + strconv.Itoa(i),
+			})
+		}
+	}
+	container.Ports = ports
```

Comment:
> This prevents us from reusing libvirtd's port allocation. IIUIC libvirtd will tell qemu which port to bind to.

Libvirt is not running inside the network namespace of the container and can therefore not decide which ports are free to use. Right now we are doing it and not libvirt, but from outside it is still "automatic".
We also have to make sure that address 0.0.0.0 is used, otherwise we can't reach the spice connection. Letting users in the future specify something else here is pretty pointless.

However, my final picture is that this all lives in virt-handler, so that a user does not see all this in the spec. I have been thinking about two ways of doing this:

 1) either get the ports then via the Status field which is not yet fully populated or
 2) directly asking virt-handler for the spec details from the node via the subresource (like with logs, exec, ... in kubectl)

> We might not need to open the port considering:

Right we could not do that, but I like that we see it. Every bit of information describing how the system is (and keeping all the bits synchronized) is IMHO pretty nice to have.

But if we move all that to virt-handler, it is very nice that we don't have to (because these fields can't be updated and the pod is already running when virt-handler comes into the picture).